[do not merge] Preparation for LLD stabilization

Kobzol opened the pull request:
This PR serves for testing test changes for stabilizing LLD.

CC @lqd

r? @ghost

try-job: dist-x86_64-linux
try-job: aarch64-gnu
try-job: aarch64-gnu-debug
try-job: x86_64-msvc-1
try-job: x86_64-msvc-2
try-job: x86_64-gnu
try-job: x86_64-gnu-stable
try-job: x86_64-gnu-aux
try-job: x86_64-gnu-debug
try-job: x86_64-apple-1
try-job: x86_64-apple-2
try-job: aarch64-apple

lqd committed:
use LLD by default on x64 regardless of channel

lqd committed:
add post-dist test for checking that we use LLD

And remove the previous beta/stable/nightly LLD tests.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13917554086/job/38943321526) [(plain)](https://github.com/rust-lang/rust/commit/549454f1302eaec7fae8b10b4db409b2b873ebda/checks/38943321526/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.139 Building wheels for collected packages: reuse
#12 3.140   Building wheel for reuse (pyproject.toml): started
#12 3.356   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.357   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.357   Stored in directory: /tmp/pip-ephem-wheel-cache-5p3c06au/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.359 Successfully built reuse
#12 3.360 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.779 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.779 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.345 Collecting virtualenv
#12 4.404   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.484      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 55.9 MB/s eta 0:00:00
#12 4.553 Collecting filelock<4,>=3.12.2
#12 4.563   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.605 Collecting platformdirs<5,>=3.9.1
#12 4.616   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.642 Collecting distlib<1,>=0.3.7
#12 4.653   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.661      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 78.3 MB/s eta 0:00:00
#12 4.742 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.933 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.6 virtualenv-20.29.3
#12 4.934 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.0s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      139200 kB
DirectMap2M:     7200768 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.224
---
Diff in /checkout/src/tools/run-make-support/src/linker.rs:1:
-use crate::{Rustc, is_msvc};
 use regex::Regex;
+
+use crate::{Rustc, is_msvc};
 
 /// Asserts that `rustc` uses LLD for linking when executed.
 pub fn assert_rustc_uses_lld(rustc: &mut Rustc) {
fmt: checked 5918 files
Build completed unsuccessfully in 0:01:19
  local time: Tue Mar 18 07:29:23 UTC 2025
  network time: Tue, 18 Mar 2025 07:29:24 GMT
##[error]Process completed with exit code 1.
```

</details>

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit 80cbbeb5fa71df8ed07988902baaf779dec43742 with merge 8a4821f34179797445396b9696f039281dd55f84...
<!-- homu: {"type":"TryBuildStarted","head_sha":"80cbbeb5fa71df8ed07988902baaf779dec43742","merge_sha":"8a4821f34179797445396b9696f039281dd55f84"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13933900334/job/39003283563)
Build commit: 8a4821f34179797445396b9696f039281dd55f84 (`8a4821f34179797445396b9696f039281dd55f84`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13933900334/job/39003283563"},"merge_sha":"8a4821f34179797445396b9696f039281dd55f84"} -->

lqd committed:
stabilize `-Zlinker-features` as `-Clinker-features`

lqd committed:
check that `-Clinker-features=[-+]lld` can only be used on x64 linux without `-Zunstable-options`

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13951839674/job/39053201779) [(plain)](https://github.com/rust-lang/rust/commit/fd52515e93b78c6d89446af1be099dc982714497/checks/39053201779/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.800 Building wheels for collected packages: reuse
#12 2.801   Building wheel for reuse (pyproject.toml): started
#12 3.021   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.022   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.022   Stored in directory: /tmp/pip-ephem-wheel-cache-ip48vjt9/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.024 Successfully built reuse
#12 3.025 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.429 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.430 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 3.984 Collecting virtualenv
#12 4.024   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.082      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 77.6 MB/s eta 0:00:00
#12 4.122 Collecting distlib<1,>=0.3.7
#12 4.125   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.133      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 89.1 MB/s eta 0:00:00
#12 4.166 Collecting platformdirs<5,>=3.9.1
#12 4.170   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.206 Collecting filelock<4,>=3.12.2
#12 4.210   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.293 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.479 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.6 virtualenv-20.29.3
#12 4.480 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.6s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      124864 kB
DirectMap2M:     9312256 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.219
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 5923 files
tidy check
/checkout/tests/ui/linking/linker-features-unstable.rs: revision [unspecified] should specify `needs-llvm-components:` as it has `--target` set
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:50
  local time: Wed Mar 19 16:45:19 UTC 2025
  network time: Wed, 19 Mar 2025 16:45:19 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lqd commented on compiler/rustc_session/src/config.rs:
```diff
@@ -2532,6 +2532,19 @@ pub fn build_session_options(early_dcx: &mut EarlyDiagCtxt, matches: &getopts::M
+    if !nightly_options::is_unstable_enabled(matches) {
+        // Check that -Clinker-features=[-+]lld is not used anywhere else than on x64
+        // without -Zunstable-options.
+        if !cg.linker_features.disabled.is_empty() || !cg.linker_features.enabled.is_empty() {
```
Let's actually check for lld instead of `!is_empty()`, I was thinking of something like `.contains(LinkerFeatures::LLD)` but then it makes it harder to check for `LinkerFeatures::CC` (to ensure it still needs `-Zunstable-options`, otherwise I think this is stabilizing +/-cc) with the 3 levels of nested negations -- so I suggest trying to instead copy the stable/unstable setup for LinkSelfContained whose check is 50 lines above, with `are_unstable_variants_set` (taking the target as parameter in our case). I think it should be easier that way. 

We can also adjust the error message below to that effect.

Kobzol replied:
I wanted to have a function like `are_unstable_variants_set`, I just thought that it's a bit weird that some remote function would be doing the check, and then the error message is generated here in `config.rs`, seems easy to be desynced (but not a big concern).

lqd replied:
The remote function wouldn't be doing the check, only returning whether something is stable or not; and the error in config is just checking that something unstable isn't called on stable? Do you have an example of a desync?

lqd replied:
Marking as resolved as we've discussing this entire PR on zulip for a while 👍

lqd committed:
stabilize `-Clink-self-contained=[+-]linker`

lqd committed:
add test for unstable values of `-Clink-self-contained`

lqd committed:
update documentation

bors commented:
:umbrella: The latest upstream changes (presumably #138865) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14243477936/job/39918658674) [(plain)](https://github.com/rust-lang/rust/commit/2cb937dde37ab18101834a0b41cead49c16c60fc/checks/39918658674/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.7s done
#19 DONE 33.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc: In function ‘bool find_and_merge_options(int, off_t, const char*, vec<cl_decoded_option>, bool, vec<cl_decoded_option>*, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc:1165:8: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 1165 |   read (fd, data, length);
      |   ~~~~~^~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
failures:

---- [ui] tests/ui/linking/link-self-contained-unstable.rs#crto stdout ----

error in revision `crto`: /checkout/tests/ui/linking/link-self-contained-unstable.rs:?: unexpected error: '?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values'

error in revision `crto`: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/link-self-contained-unstable.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "crto" "--check-cfg" "cfg(test,FALSE,crto,libc,unwind,sanitizers,mingw)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-unstable.crto" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Clink-self-contained=+crto"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values
---

thread '[ui] tests/ui/linking/link-self-contained-unstable.rs#crto' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [ui] tests/ui/linking/link-self-contained-unstable.rs#mingw stdout ----

error in revision `mingw`: /checkout/tests/ui/linking/link-self-contained-unstable.rs:?: unexpected error: '?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values'

error in revision `mingw`: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/link-self-contained-unstable.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "mingw" "--check-cfg" "cfg(test,FALSE,crto,libc,unwind,sanitizers,mingw)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-unstable.mingw" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Clink-self-contained=+mingw"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values
---

thread '[ui] tests/ui/linking/link-self-contained-unstable.rs#mingw' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected

---- [ui] tests/ui/linking/link-self-contained-unstable.rs#libc stdout ----

error in revision `libc`: /checkout/tests/ui/linking/link-self-contained-unstable.rs:?: unexpected error: '?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values'

error in revision `libc`: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/link-self-contained-unstable.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "libc" "--check-cfg" "cfg(test,FALSE,crto,libc,unwind,sanitizers,mingw)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-unstable.libc" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Clink-self-contained=-libc"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values
---

thread '[ui] tests/ui/linking/link-self-contained-unstable.rs#libc' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected

---- [ui] tests/ui/linking/linker-features-unstable-cc.rs stdout ----

error: /checkout/tests/ui/linking/linker-features-unstable-cc.rs:?: unexpected error: '?:?: ?:?: incorrect value `+cc` for codegen option `linker-features` - a list of enabled (`+` prefix) and disabled (`-` prefix) features: `lld` was expected'

error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/linker-features-unstable-cc.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/linker-features-unstable-cc" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target=x86_64-unknown-linux-gnu" "-C" "linker-features=+cc" "--crate-type=rlib"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: incorrect value `+cc` for codegen option `linker-features` - a list of enabled (`+` prefix) and disabled (`-` prefix) features: `lld` was expected
---

thread '[ui] tests/ui/linking/linker-features-unstable-cc.rs' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected

---- [ui] tests/ui/linking/linker-features-lld-disallowed-target.rs stdout ----

error: /checkout/tests/ui/linking/linker-features-lld-disallowed-target.rs:?: unexpected error: '?:?: ?:?: `-C linker-features` with lld are unstable for the `x86_64-unknown-linux-musl target, ` the `-Z unstable-options` flag must also be passed to use it on this target'

error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/linker-features-lld-disallowed-target.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/linker-features-lld-disallowed-target" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target=x86_64-unknown-linux-musl" "-C" "linker-features=-lld" "--crate-type=rlib"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: `-C linker-features` with lld are unstable for the `x86_64-unknown-linux-musl target, ` the `-Z unstable-options` flag must also be passed to use it on this target
---

thread '[ui] tests/ui/linking/linker-features-lld-disallowed-target.rs' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected

---- [ui] tests/ui/linking/link-self-contained-unstable.rs#unwind stdout ----

error in revision `unwind`: /checkout/tests/ui/linking/link-self-contained-unstable.rs:?: unexpected error: '?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values'

error in revision `unwind`: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/link-self-contained-unstable.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "unwind" "--check-cfg" "cfg(test,FALSE,crto,libc,unwind,sanitizers,mingw)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-unstable.unwind" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Clink-self-contained=+unwind"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values
---

thread '[ui] tests/ui/linking/link-self-contained-unstable.rs#unwind' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected

---- [ui] tests/ui/linking/link-self-contained-unstable.rs#sanitizers stdout ----

error in revision `sanitizers`: /checkout/tests/ui/linking/link-self-contained-unstable.rs:?: unexpected error: '?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values'

error in revision `sanitizers`: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/link-self-contained-unstable.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "sanitizers" "--check-cfg" "cfg(test,FALSE,crto,libc,unwind,sanitizers,mingw)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-unstable.sanitizers" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Clink-self-contained=-sanitizers"

--- unexpected errors (from JSON output) ---
ERROR     line   ?: ?:?: ?:?: only `-C link-self-contained` values `y`/`yes`/`on`/`n`/`no`/`off`/`-linker`/`+linker` are stable, the `-Z unstable-options` flag must also be passed to use the unstable values
---

thread '[ui] tests/ui/linking/link-self-contained-unstable.rs#sanitizers' panicked at src/tools/compiletest/src/runtest.rs:799:13:
errors differ from expected

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #139354) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd committed:
update `check_unstable_variants`'s handling of `-C linker-features=[-+]lld`

- separate enabling and disabling the feature in the error
- add both polarities to the dedicated test
- update documentation and precondition

Kobzol commented on src/doc/rustc/src/codegen-options/index.md:
```diff
@@ -244,18 +244,35 @@ flag for another way to specify the linker.
-Currently, only one such feature is stable:
-- `lld`: toggles the usage of the lld linker, either the system-installed binary, or the self-contained
-  `rust-lld` linker.
+The flag accepts a comma-separated list of features, individually enabled (`+features`) or disabled
```
I think that an example would be nice. It's a bit confusing that it says +features rather than +feature, and after reading this, I wouldn't be sure how to actually use it.

lqd replied:
Ah good catch, the `+features` is a typo from the unstable documentation. I've fixed it and added examples.

Kobzol commented on src/doc/rustc/src/codegen-options/index.md:
```diff
@@ -244,18 +244,35 @@ flag for another way to specify the linker.
+To opt out of using the `lld` linker (for targets where it is configured as the default linker),
 use `-Clinker-features=-lld`.
 
+In general, this design allows:
```
This sounds a bit more like an internal comment than user-facing documentation.

lqd replied:
It's the unstable book documentation but ok I've removed it.

lqd commented:
Might as well @bors try

bors commented:
:hourglass: Trying commit 49cabaabb6e50903f7f75e1f490430192ccbe0d4 with merge 0c623eb581ae54ed553869852e1d1b3bf0bb6a82...
<!-- homu: {"type":"TryBuildStarted","head_sha":"49cabaabb6e50903f7f75e1f490430192ccbe0d4","merge_sha":"0c623eb581ae54ed553869852e1d1b3bf0bb6a82"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14273583451/job/40017761192)
Build commit: 0c623eb581ae54ed553869852e1d1b3bf0bb6a82 (`0c623eb581ae54ed553869852e1d1b3bf0bb6a82`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14273583451/job/40017761192"},"merge_sha":"0c623eb581ae54ed553869852e1d1b3bf0bb6a82"} -->

lqd committed:
update bootstrap mcp510 handling

now that it's been stabilized, beta and stage1 need to use different
flags (-C vs -Z)

lqd commented:
@bors try

bors commented:
:hourglass: Trying commit 5d920a34c1533a6d6cbee16782b499d7599c9300 with merge f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"5d920a34c1533a6d6cbee16782b499d7599c9300","merge_sha":"f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/14668610617/job/41169592571) [(plain)](https://github.com/rust-lang-ci/rust/commit/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/checks/41169592571/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
##[group]Building compiler artifacts (stage1 -> stage2, x86_64-unknown-linux-gnu)
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg -Wwarnings` (exit status: 1)
  --- stderr
  error: unknown unstable option: `linker-features`

  [RUSTC-TIMING] ___ test:false 0.010
Build completed unsuccessfully in 0:07:50
---
Caused by:
    Command RUST_BACKTRACE=full python3 /checkout/x.py build --target x86_64-unknown-linux-gnu --host x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate /tmp/tmp-multistage/opt-artifacts/rustc-pgo --set llvm.thin-lto=false --set llvm.link-shared=true [at /checkout/obj] has failed with exit code Some(1)

Stack backtrace:
   0: <anyhow::Error>::msg::<alloc::string::String>
             at /rust/deps/anyhow-1.0.97/src/backtrace.rs:27:14
   1: <opt_dist::exec::CmdBuilder>::run
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/exec.rs:80:17
   2: <opt_dist::exec::Bootstrap>::run
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/exec.rs:181:9
   3: opt_dist::execute_pipeline::{closure#1}::{closure#0}
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/main.rs:230:13
   4: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}::{closure#0}, ()>
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/timer.rs:111:22
   5: opt_dist::execute_pipeline::{closure#1}
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/main.rs:219:9
   6: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}, opt_dist::training::RustcPGOProfile>
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/timer.rs:111:22
   7: opt_dist::execute_pipeline
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/main.rs:216:29
   8: opt_dist::main
             at /rustc/f88e20107e2356d368ffcc8c3e225e7e7e1f6d4a/src/tools/opt-dist/src/main.rs:416:18
   9: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/core/src/ops/function.rs:250:5
  10: std::sys::backtrace::__rust_begin_short_backtrace::<fn() -> core::result::Result<(), anyhow::Error>, core::result::Result<(), anyhow::Error>>
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/sys/backtrace.rs:152:18
  11: std::rt::lang_start::<core::result::Result<(), anyhow::Error>>::{closure#0}
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/rt.rs:199:18
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/core/src/ops/function.rs:284:13
  13: std::panicking::try::do_call
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/panicking.rs:589:40
  14: std::panicking::try
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/panicking.rs:552:19
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14668610617/job/41170603652)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/14668610617/job/41170603652","builder_name":"checks-actions"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14704713919/job/41270955843)
Build commit: 13c5d06e969c464d34897752700e82d5f7689255 (`13c5d06e969c464d34897752700e82d5f7689255`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14704713919/job/41270955843"},"merge_sha":"13c5d06e969c464d34897752700e82d5f7689255"} -->

Kobzol commented:
@rust-timer build 13c5d06e969c464d34897752700e82d5f7689255

rust-timer commented:
Queued 13c5d06e969c464d34897752700e82d5f7689255 with parent 0134651fb81314870903e21b1bcbdd993d75b61a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0134651fb81314870903e21b1bcbdd993d75b61a&end=13c5d06e969c464d34897752700e82d5f7689255).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit d8447297f53ea6bfeb1a95e2b775d2a3c1643886 with merge 27d41fa9d39da85acd968a3fc380c3d7266224ba...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d8447297f53ea6bfeb1a95e2b775d2a3c1643886","merge_sha":"27d41fa9d39da85acd968a3fc380c3d7266224ba"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14708491374/job/41286554621)
Build commit: 27d41fa9d39da85acd968a3fc380c3d7266224ba (`27d41fa9d39da85acd968a3fc380c3d7266224ba`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14708491374/job/41286554621"},"merge_sha":"27d41fa9d39da85acd968a3fc380c3d7266224ba"} -->

rust-timer commented:
Finished benchmarking commit (13c5d06e969c464d34897752700e82d5f7689255): [comparison URL](https://perf.rust-lang.org/compare.html?start=0134651fb81314870903e21b1bcbdd993d75b61a&end=13c5d06e969c464d34897752700e82d5f7689255&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=0134651fb81314870903e21b1bcbdd993d75b61a&end=13c5d06e969c464d34897752700e82d5f7689255&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0%  | [3.0%, 3.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.4%, -0.4%] | 1     |
| All ❌✅ (primary)                 | 3.0%  | [3.0%, 3.0%]   | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0134651fb81314870903e21b1bcbdd993d75b61a&end=13c5d06e969c464d34897752700e82d5f7689255&stat=max-rss)
<details>
<summary>Results (primary 0.7%, secondary 3.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.4%, 1.5%] | 12    |
| Regressions ❌ <br /> (secondary)  | 3.3% | [3.3%, 3.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.4%, 1.5%] | 12    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0134651fb81314870903e21b1bcbdd993d75b61a&end=13c5d06e969c464d34897752700e82d5f7689255&stat=cycles:u)
<details>
<summary>Results (primary 1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8% | [0.6%, 3.1%] | 2     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.8% | [0.6%, 3.1%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=0134651fb81314870903e21b1bcbdd993d75b61a&end=13c5d06e969c464d34897752700e82d5f7689255&stat=size:linked_artifact)
<details>
<summary>Results (primary 1.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [1.1%, 1.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [1.1%, 1.1%] | 1     |
</details>

**Bootstrap**: 761.588s -> 762.995s (0.18%)
**Artifact size**: 365.21 MiB -> 365.19 MiB (-0.01%)

lqd commented:
perf results look like noise

Kobzol commented:
@bors try

lqd commented:
🚀

lqd committed:
simplify test annotations

lqd committed:
check that `-Clink-self-contained=[-+]linker` can only be used on x64 linux without `-Zunstable-options`

lqd commented:
@bors try

bors commented:
:hourglass: Trying commit b92dfd0adbc8ab74841a5565910a11074b7b094a with merge 9365844dc5ae2940b1dffede2b3f91612b041b38...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b92dfd0adbc8ab74841a5565910a11074b7b094a","merge_sha":"9365844dc5ae2940b1dffede2b3f91612b041b38"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14842144756/job/41669614505)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/14842144756/job/41669614505","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14842073628/job/41667278928) [(plain)](https://github.com/rust-lang/rust/commit/b92dfd0adbc8ab74841a5565910a11074b7b094a/checks/41667278928/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.1s done
#18 DONE 33.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 5.572
[RUSTC-TIMING] test test:false 9.359
    Finished `release` profile [optimized] target(s) in 1m 18s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 78.215
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.774
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.922
    Finished `release` profile [optimized] target(s) in 4.02s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.203
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.181
---

failures:

---- [ui] tests/ui/linking/link-self-contained-consistency.rs#one stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-consistency.one/link-self-contained-consistency.one.stderr"
diff of stderr:

- error: some `-C link-self-contained` components were both enabled and disabled: linker
+ error: `-C link-self-contained=+linker` is unstable on the `i686-unknown-linux-gnu` target. The `-Z unstable-options` flag must also be passed to use it on this target
2 
3 


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args linking/link-self-contained-consistency.rs`

error in revision `one`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/linking/link-self-contained-consistency.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--cfg" "one" "--check-cfg" "cfg(test,FALSE,one,many)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/linking/link-self-contained-consistency.one" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=x86_64-linux-gnu-gcc" "-Clink-self-contained=-linker" "-Clink-self-contained=+linker"
stdout: none
--- stderr -------------------------------
error: `-C link-self-contained=+linker` is unstable on the `i686-unknown-linux-gnu` target. The `-Z unstable-options` flag must also be passed to use it on this target
------------------------------------------



failures:
```

</details>

lqd commented:
@bors try

bors commented:
:hourglass: Trying commit 2b829ba23376c02350f0ea5e6a0f7f06454e009d with merge dd3d9ce1de842e3c608e8c4c6764cd700df0becd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2b829ba23376c02350f0ea5e6a0f7f06454e009d","merge_sha":"dd3d9ce1de842e3c608e8c4c6764cd700df0becd"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14843261535/job/41680683223)
Build commit: dd3d9ce1de842e3c608e8c4c6764cd700df0becd (`dd3d9ce1de842e3c608e8c4c6764cd700df0becd`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14843261535/job/41680683223"},"merge_sha":"dd3d9ce1de842e3c608e8c4c6764cd700df0becd"} -->
